### PR TITLE
[Merged by Bors] - port the clean tactic from lean 3 to lean 4

### PR DESCRIPTION
**Pull request by mkaratarakis**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/Clean.lean [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+   | _ => none)
+
+end Lean.Expr
+
+/--
+Remove identity functions from a term. These are normally
+automatically generated with terms like `show t, from p` or
+`(p : t)` which translate to some variant on `@id t p` in
+order to retain the ty/-! # `classical` and `classical!` tactics -/
+pe.
```
Comment (alexjbest) [inline Mathlib/Tactic/Clean.lean:42]:
```suggestion
order to retain the type.
```

**Comment by kmill** [comment]
I'd suggest that once #5717 is merged that we put this tactic in the DefeqTransformations file and use the `runDefeqTactic` function to run it, like the other tactics there. That means you don't have to reinvent how to run an expression-transformation tactic, and you get `clean at h` and `clean at *` for free.

**Comment by kmill** [comment]
Ah, I really misunderstood what this tactic was about. Now I understand that `clean t` is like `exact t` but removes type hints from `t`, so my suggestion to use the defeq transformations framework doesn't make sense.

@digama0 You write the original version of `clean`. It looks like it was only ever used in a single place in mathlib -- is it still useful? Should we turn it into a `cleanHints%` term elaborator instead? That way you can write stuff like `let x := cleanHints% by ...` or `exact cleanHints% by ...`. Or maybe it could be a `classical`-like tactic that takes a tactic sequence instead of a term?

**Comment by digama0** [comment]
I think it would make sense as a term elaborator.

File: Mathlib/Tactic/Clean.lean [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+open Lean Meta Elab
+
+namespace Lean.Expr
+
+/-- List of names removed by the `clean` tactic.
+All of these names must resolve to functions defeq `id`. -/
+-- Note: one could also add `hidden`, but this doesn't arise from type hints.
+def cleanConsts : List Name :=
+  [``id, ``Id]
```
Comment (kmill) [inline Mathlib/Tactic/Clean.lean:23]:
```suggestion
  [``id]
```
`Id` takes only one argument, not two

**Comment by PatrickMassot** [comment]
bors merge

**Pull request closed without merging**